allennlp/models/vilbert.py

```diff
@@ -0,0 +1,836 @@
+        embedding_output = embedding_output.unsqueeze(1).expand(-1, 2, -1, -1)
+        attention_mask = attention_mask.unsqueeze(1).expand(-1, 2, -1)
+
+        # We create a 3D attention mask from a 2D tensor mask.
+        # Sizes are [batch_size, 1, 1, to_seq_length]
+        # So we can broadcast to [batch_size, num_heads, from_seq_length, to_seq_length]
+        # this attention mask is more simple than the triangular masking of
+        # causal attention used in OpenAI GPT, we just need to prepare the
+        # broadcast dimension here.
+        extended_attention_mask = attention_mask.unsqueeze(2).unsqueeze(3)
+        extended_image_attention_mask = image_attention_mask.unsqueeze(2).unsqueeze(3)
+
+        extended_attention_mask2 = attention_mask.unsqueeze(3)
+
+        # Since attention_mask is 1.0 for positions we want to attend and 0.0 for
+        # masked positions, this operation will create a tensor which is 0.0 for
+        # positions we want to attend and -10000.0 for masked positions.
+        # Since we are adding it to the raw scores before the softmax, this is
+        # effectively the same as removing these entirely.
+        extended_attention_mask = extended_attention_mask.to(
+            dtype=next(self.parameters()).dtype
+        )  # fp16 compatibility
+        extended_attention_mask = (1.0 - extended_attention_mask) * -10000.0
+
+        extended_attention_mask2 = extended_attention_mask2.to(
+            dtype=next(self.parameters()).dtype
+        )  # fp16 compatibility
+
+        extended_image_attention_mask = extended_image_attention_mask.to(
+            dtype=next(self.parameters()).dtype
+        )  # fp16 compatibility
+        extended_image_attention_mask = (1.0 - extended_image_attention_mask) * -10000.0
+
+        co_attention_mask = torch.zeros(batch_size, num_images, feature_size, num_tokens).type_as(
+            extended_image_attention_mask
+        )
+
+        extended_co_attention_mask = co_attention_mask.unsqueeze(2)
+
+        # (batch_size, num_images, num_boxes, image_embedding_dim)
+        v_embedding_output = self.image_embeddings(visual_features, box_coordinates)
+        encoded_layers_t, encoded_layers_v = self.encoder(
+            embedding_output,
+            v_embedding_output,
+            extended_attention_mask,
+            extended_attention_mask2,
+            extended_image_attention_mask,
+            extended_co_attention_mask,
+        )
+
+        sequence_output_t = encoded_layers_t[:, :, :, :, -1]
+        sequence_output_v = encoded_layers_v[:, :, :, :, -1]
+
+        pooled_output_t = self.t_pooler(sequence_output_t)
+        pooled_output_v = self.v_pooler(sequence_output_v)
+
+        if self.fusion_method == "sum":
+            pooled_output = self.dropout(pooled_output_t + pooled_output_v)
+        elif self.fusion_method == "mul":
+            pooled_output = self.dropout(pooled_output_t * pooled_output_v)
+
```

Comment:
This is missing an `else: raise ValueError(...)`.

Reply:
Done.